sopel_modules/github/webhook.py

```diff
@@ -124,13 +129,75 @@ def process_payload(payload, targets):
             sopel_instance.say(message, row[0])
 
 
+def debug_log_request(request_headers, request_body):
+    LOGGER.debug('Headers: {}'.format(dict([(k, request_headers[k]) for k in request_headers])))
+    LOGGER.debug('Request: {}'.format(request_body.decode('utf-8')))
+
+
+def abort_request(status_code=400, response_message=None):
+    if sopel_instance.config.github.debug_mode:
+        LOGGER.warning('`debug_mode = True`; allowing unverified request...')
+        return None
+    return bottle.abort(status_code, response_message)
+
+
+def verify_request():
+    request_headers = bottle.request.headers
+    request_body = bottle.request.body.read()
+
+    if not request_headers.get('X-Hub-Signature'):
+        msg = 'Request is missing a hash signature.'
+        LOGGER.error(msg)
+        debug_log_request(request_headers, request_body)
+        return abort_request(401, msg)  # 401 Unauthorized; missing required header
+
+    digest_name, payload_signature = request_headers.get('X-Hub-Signature').split('=')
+    # Currently, GitHub only uses 'sha1'; log a warning if a different digest is
+    # specified by the server. GitHub may have started using a new digest and
+    # this should be confirmed.
+    if digest_name != 'sha1':
+        LOGGER.warning('Unexpected signature digest: {}'.format(digest_name))
+        debug_log_request(request_headers, request_body)
+
+    try:
+        digest_mod = getattr(hashlib, digest_name)
+    except AttributeError:
+        # The previous digest check does not require a 'sha1' digest, but simply
+        # warns when an unexpected digest is specified. The function will
+        # attempt to find the digest specified in the signature, but if it is
+        # not currently supported by Python's `hashlib`, an error will be logged
+        # and returned.
+        msg = 'Unsupported signature digest: {}'.format(digest_name)
+        LOGGER.error(msg)
+        debug_log_request(request_headers, request_body)
+        return abort_request(501, msg)  # 501 Not Implemented; server does not support the functionality required to fulfill the request
+
+    secret = sopel_instance.config.github.webhook_secret
+    hash_ = hmac.new(secret.encode('utf-8') if secret else None, msg=request_body, digestmod=digest_mod)
+    expected_signature = hash_.hexdigest()
+    if payload_signature != expected_signature:
```

Comment:
Should use [`hmac.compare_digest()`](https://docs.python.org/3/library/hmac.html#hmac.compare_digest) to mitigate timing attacks, no?

Reply:
```suggestion
    if not hmac.compare_digest(payload_signature, expected_signature):
```